File: src/restricted-extras-ktextaddons/libconverttext.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const QTextCursor = @import("libqt6").QTextCursor;
const std = @import("std");

/// ### [Upstream resources](https://api.kde.org/legacy/ktextaddons/html/classTextUtils_1_1ConvertText.html)
pub const TextUtils__ConvertText = extern struct {
    /// ### [Upstream resources](https://api.kde.org/legacy/ktextaddons/html/classTextUtils_1_1ConvertText.html)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` str: []const u8 `
    ///
    pub fn Normalize(allocator: std.mem.Allocator, str: []const u8) []const u8 {
        const str_str = qtc.libqt_string{
            .len = str.len,
            .data = str.ptr,
        };
        var _str = qtc.TextUtils__ConvertText_Normalize(str_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("textutils__converttext.Normalize: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### [Upstream resources](https://api.kde.org/legacy/ktextaddons/html/classTextUtils_1_1ConvertText.html)
    ///
    /// ## Parameter(s):
    ///
    /// ` cursor: QTextCursor `
    ///
    pub fn UpperCase(cursor: anytype) void {
        comptime _ = @TypeOf(cursor)._is_QTextCursor;
        qtc.TextUtils__ConvertText_UpperCase(@ptrCast(cursor.ptr));
    }

    /// ### [Upstream resources](https://api.kde.org/legacy/ktextaddons/html/classTextUtils_1_1ConvertText.html)
    ///
    /// ## Parameter(s):
    ///
    /// ` cursor: QTextCursor `
    ///
    pub fn LowerCase(cursor: anytype) void {
        comptime _ = @TypeOf(cursor)._is_QTextCursor;
        qtc.TextUtils__ConvertText_LowerCase(@ptrCast(cursor.ptr));
    }

    /// ### [Upstream resources](https://api.kde.org/legacy/ktextaddons/html/classTextUtils_1_1ConvertText.html)
    ///
    /// ## Parameter(s):
    ///
    /// ` cursor: QTextCursor `
    ///
    pub fn SentenceCase(cursor: anytype) void {
        comptime _ = @TypeOf(cursor)._is_QTextCursor;
        qtc.TextUtils__ConvertText_SentenceCase(@ptrCast(cursor.ptr));
    }

    /// ### [Upstream resources](https://api.kde.org/legacy/ktextaddons/html/classTextUtils_1_1ConvertText.html)
    ///
    /// ## Parameter(s):
    ///
    /// ` cursor: QTextCursor `
    ///
    pub fn ReverseCase(cursor: anytype) void {
        comptime _ = @TypeOf(cursor)._is_QTextCursor;
        qtc.TextUtils__ConvertText_ReverseCase(@ptrCast(cursor.ptr));
    }
};

File: src/restricted-extras-ktextaddons/libconverttext.cpp
#include <QString>
#include <QTextCursor>
#define WORKAROUND_INNER_CLASS_DEFINITION_TextUtils__ConvertText
#include <converttext.h>
#include "libconverttext.h"
#include "libconverttext.hxx"

libqt_string TextUtils__ConvertText_Normalize(libqt_string str) {
    QString str_QString = QString::fromUtf8(str.data, str.len);
    auto _ret = TextUtils::ConvertText::normalize(str_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void TextUtils__ConvertText_UpperCase(QTextCursor* cursor) {
    TextUtils::ConvertText::upperCase(*cursor);
}

void TextUtils__ConvertText_LowerCase(QTextCursor* cursor) {
    TextUtils::ConvertText::lowerCase(*cursor);
}

void TextUtils__ConvertText_SentenceCase(QTextCursor* cursor) {
    TextUtils::ConvertText::sentenceCase(*cursor);
}

void TextUtils__ConvertText_ReverseCase(QTextCursor* cursor) {
    TextUtils::ConvertText::reverseCase(*cursor);
}
